### PR TITLE
Add a autocomplete field to team choosing screen

### PR DESCRIPTION
**Pull request by vanessa**
## Description

https://trello.com/c/KiPfOfFV/5-as-a-trainer-i-can-see-available-pkns-as-i-type-their-names-in-the-form-type-ahead

## Motivation and Context

It gets boring for users to always input the Pokemon id and have them memorized.

So, now that it's possible to search the Pokemon directly from the db (#9), a typeahead field was added to make the UI better and improve UX.

## Screenshots (if appropriate):
choosing

## Critical points affected:

- [ ] Signup flow
- [ ] Battle creation
- [x] Battle team choosing
- [ ] Battle execution

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change which improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt (refactoring existing code)

## Checklist:

- [ ] My change requires documentation updates.
- [ ] I have updated the documentation accordingly.
- [ ] My change requires dependencies updates.
- [ ] I have updated the dependencies accordingly.
- [x] My change affected critical flows & code.
- [x] I have tested the critical flows & code accordingly.
- [ ] My change requires environment variables updates.
- [ ] I have described the environment variables accordingly.

### CODE REVIEW

File: templates/battles/choose_team.html [hunk 1]
```diff
@@ -12,7 +13,6 @@ <h2>Choose your Pokemons!</h2>
         </div>
         {% if form.errors %}
             <div class="form-error">{{ form.non_field_errors }}</div>
-            <div class="form-error">{{ form.errors }}</div>
```
Comment (hugobessa) [inline templates/battles/choose_team.html]: why was this removed?

File: battles/views.py [hunk 2]
```diff
@@ -67,6 +70,25 @@ def get_context_data(self, **kwargs):
         return context
 
 
+class PokemonAutocompleteView(autocomplete.Select2QuerySetView):
```
Comment (hugobessa) [inline battles/views.py]: A better name would be PokemonListAPIView

File: battles/urls.py [hunk 3]
```diff
@@ -1,13 +1,15 @@
...
 
 urlpatterns = [
     url(r'^$', BattlesListView.as_view(), name='list'),
 
     url(r'^create$', CreateBattleView.as_view(), name='create-battle'),
     url(r'^details/(?P<pk>[\w-]+)$', BattleView.as_view(), name='details'),
-    url(r'^details/(?P<pk>[\w-]+)/team$',
-        ChoosePokemonTeamView.as_view(), name='team')
+    url(r'^details/(?P<pk>[\w-]+)/team$', ChoosePokemonTeamView.as_view(), name='team'),
+    url(r'^pokemon-autocomplete$', PokemonAutocompleteView.as_view(), name='pokemon-autocomplete')
```
Comment (hugobessa) [inline battles/urls.py]: maybe a better url for this endpoint would be '^api/pokemon/$', and the name api-pokemon-list. This library implements a RESTful list api, so it would be better if it looked more like that

File: battles/forms.py [hunk 4]
```diff
@@ -24,29 +26,46 @@ class Meta:
...
+            attrs={
+                'data-placeholder': 'Select the second Pokemon',
+                'data-html': True
+            })
+    )
+
+    third_pokemon = forms.ModelChoiceField(
+        queryset=Pokemon.objects.all(),
+        widget=autocomplete.ModelSelect2(
+            url='/battles/pokemon-autocomplete',
```
Comment (hugobessa) [inline battles/forms.py]: check if is it possible to use [reverse_lazy](https://docs.djangoproject.com/en/2.0/ref/urlresolvers/#reverse-lazy) here

File: battles/forms.py [hunk 5]
```diff
@@ -24,29 +26,46 @@ class Meta:
...
+            attrs={
+                'data-placeholder': 'Select the first Pokemon',
+                'data-html': True
+            })
+    )
+
+    second_pokemon = forms.ModelChoiceField(
+        queryset=Pokemon.objects.all(),
+        widget=autocomplete.ModelSelect2(
+            url='/battles/pokemon-autocomplete',
```
Comment (hugobessa) [inline battles/forms.py]: same as bellow

File: battles/forms.py [hunk 6]
```diff
@@ -24,29 +26,46 @@ class Meta:
...
-    second_pokemon = forms.IntegerField(
-        min_value=1, max_value=802, required=True, label='Second pokemon',
-        validators=[validate_integer_is_not_zero])
-    third_pokemon = forms.IntegerField(
-        min_value=1, max_value=802, required=True, label='Third pokemon',
-        validators=[validate_integer_is_not_zero])
+    first_pokemon = forms.ModelChoiceField(
+        queryset=Pokemon.objects.all(),
+        widget=autocomplete.ModelSelect2(
+            url='/battles/pokemon-autocomplete',
```
Comment (hugobessa) [inline battles/forms.py]: same as bellow

**Review by hugobessa — CHANGES_REQUESTED** [review]
the code is very good, @vanessa! :100: 
Just made some comments

File: templates/battles/choose_team.html (hunk 1 above)
Comment (vanessa) [reply templates/battles/choose_team.html]: Because I don't have any field related errors for this form.

Comment (hugobessa) [reply templates/battles/choose_team.html]: I think is better to maintain this indeed, because if, for some reason, you add some field related errors in the future you'll have to remember to add this again on template. Leaving this implicates in any problem?

Comment (vanessa) [reply templates/battles/choose_team.html]: True. I'm gonna add.

**Comment by vanessa** [comment]
All done :heavy_check_mark:

**Review by hugobessa — APPROVED** [review]

File: pokebattle/urls.py [hunk 7]
```diff
@@ -5,14 +5,21 @@
...
+
+    # Pokebattle
+
     url(r'^', include('users.urls', namespace='auth')),
     url(r'^$', TemplateView.as_view(template_name='common/index.html'), name='home'),
-    url(r'^battles/', include('battles.urls', namespace='battles'))
+    url(r'^battles/', include('battles.urls', namespace='battles')),
+
+    # api
+    url(r'^api/pokemon$', PokemonListAPIView.as_view(), name='api-pokemon-list'),
```
Comment (hugobessa) [inline pokebattle/urls.py]: I think you forgot the traing bar in the url

Comment (vanessa) [reply pokebattle/urls.py]: Done.

Comment (hugobessa) [inline pokebattle/urls.py]: I think you forgot the trailing bar in the url here

File: pokebattle/urls.py [hunk 8]
```diff
@@ -5,14 +5,21 @@
...
+
+    # Pokebattle
+
     url(r'^', include('users.urls', namespace='auth')),
     url(r'^$', TemplateView.as_view(template_name='common/index.html'), name='home'),
-    url(r'^battles/', include('battles.urls', namespace='battles'))
+    url(r'^battles/', include('battles.urls', namespace='battles')),
+
+    # api
+    url(r'^api/pokemon/$', PokemonListAPIView.as_view(), name='api-pokemon-list'),
```
Comment (hugobessa) [inline pokebattle/urls.py]: I think you forgot the trailing bar in the url here

File: pokebattle/urls.py (hunk 7 above)
Comment (vanessa) [reply pokebattle/urls.py]: Done.
